### PR TITLE
store: support topn in localstore

### PR DESCRIPTION
**Pull request by hanfei1991**
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu PTAL

### CODE REVIEW

File: store/localstore/local_region.go [hunk 1]
```diff
@@ -44,18 +46,131 @@ type regionResponse struct {
...
+func (t *topnSolver) Push(x interface{}) {
+	t.rows = append(t.rows, x.(*sortRow))
+	t.heapSize++
+}
+
+func (t *topnSolver) Pop() interface{} {
+	return nil
+}
+
+func (t *topnSolver) Less(i, j int) bool {
```
Comment (XuHuaiyu) [inline store/localstore/local_region.go]: extract the dumplicated code to be a method as **less([]*tipb.ByItem)**
or
directly invoke t.topnContainer.Less() 
 may be better

File: store/localstore/local_region.go [hunk 2]
```diff
@@ -44,18 +46,131 @@ type regionResponse struct {
...
+		}
+	}
+
+	return false
+}
+
+type topnSolver struct {
+	topnContainer
+
+	totalCount int
```
Comment (XuHuaiyu) [inline store/localstore/local_region.go]: comment for totalCount & heapSize may be easier to distinguish

File: store/localstore/local_region.go [hunk 3]
```diff
@@ -83,6 +198,30 @@ func (rs *localRegion) Handle(req *regionRequest) (*regionResponse, error) {
 			ctx.whereColumns = make(map[int64]*tipb.ColumnInfo)
 			collectColumnsInExpr(sel.Where, ctx, ctx.whereColumns)
 		}
+		if len(sel.OrderBy) > 0 {
+			if sel.OrderBy[0].Expr == nil {
+				ctx.desc = sel.OrderBy[0].Desc
+			} else {
+				if sel.Limit == nil {
+					return nil, errors.New("We don't supported pushing down a sort without any limit.")
```
Comment (XuHuaiyu) [inline store/localstore/local_region.go]: supported-->support
a sort --> Sort (better?)
any limit --> Limit (better?)

File: store/localstore/local_region.go [hunk 4]
```diff
@@ -83,6 +198,30 @@ func (rs *localRegion) Handle(req *regionRequest) (*regionResponse, error) {
...
+					topnContainer: topnContainer{
+						orderByItems: sel.OrderBy,
+					},
+				}
+				ctx.topnColumns = make(map[int64]*tipb.ColumnInfo)
+				for _, item := range sel.OrderBy {
+					collectColumnsInExpr(item.Expr, ctx, ctx.topnColumns)
+				}
+				for k := range ctx.whereColumns {
+					// It is will be handled in where.
```
Comment (XuHuaiyu) [inline store/localstore/local_region.go]: It is --> It

File: store/localstore/local_region.go (hunk 1 above)
Comment (hanfei1991) [inline store/localstore/local_region.go]: I don't get your points. How to extract the dup code ? Pass a bool flag as argument?

File: store/localstore/local_region.go [hunk 5]
```diff
@@ -44,18 +46,133 @@ type regionResponse struct {
...
 	groups       map[string]bool
 	groupKeys    [][]byte
 	aggregates   []*aggregateFuncExpr
-	aggregate    bool
+	topnSolver   *topnSolver
 	keyRanges    []kv.KeyRange
 
+	// TODO: Only one of these three flags can be true at the same time. We should set this as an enum var.
+	aggregate bool
+	desc      bool
```
Comment (coocood) [inline store/localstore/local_region.go]: At the first sight, we would think `desc` and other two flags can be true at the same time.
Maybe `desc` should be renamed to something more specific?

File: store/localstore/local_region.go [hunk 6]
```diff
@@ -44,18 +46,133 @@ type regionResponse struct {
...
+			return true
+		} else if ret > 0 {
+			return false
+		}
+	}
+
+	return false
+}
+
+type topnSolver struct {
```
Comment (coocood) [inline store/localstore/local_region.go]: topnContainer implements the `sort.Interface`, why not name it more explicitly, like `topnSorter`? And topnSolver implements `Heap` interface, why not name it `topnHeap`.

Why embed topnContainer in topnSolver? 
It's convenient to write, but one doesn't belong to another conceptually.

File: store/localstore/local_region.go [hunk 7]
```diff
@@ -44,18 +46,133 @@ type regionResponse struct {
...
+			return true
+		} else if ret < 0 {
+			return false
+		}
+	}
+
+	return false
+}
+
+func (t *topnSolver) tryToAddRow(row *sortRow) bool {
```
Comment (coocood) [inline store/localstore/local_region.go]: Add comments about when `tryToAddRow` returns false.

**Comment by shenli** [comment]
Please add some comment for the code.

File: store/localstore/local_region.go (hunk 1 above)
Comment (XuHuaiyu) [inline store/localstore/local_region.go]:
``` go
func less([]*tipb.ByItem) bool {
   // duplicated code
}

func (t * topnHeap) Less(i, j int) bool {
      return less(t.orderByItems)
}
func (t *topnSorter) Less(i, j int) bool {
      return less(t.orderByItems)
}
```

Comment (hanfei1991) [inline store/localstore/local_region.go]: There are some differences bewteen these two functions......

**Comment by hanfei1991** [comment]
@coocood @shenli PTAL

**Comment by coocood** [comment]
LGTM

File: store/localstore/local_region.go [hunk 8]
```diff
@@ -83,6 +207,30 @@ func (rs *localRegion) Handle(req *regionRequest) (*regionResponse, error) {
 			ctx.whereColumns = make(map[int64]*tipb.ColumnInfo)
 			collectColumnsInExpr(sel.Where, ctx, ctx.whereColumns)
 		}
+		if len(sel.OrderBy) > 0 {
+			if sel.OrderBy[0].Expr == nil {
```
Comment (shenli) [inline store/localstore/local_region.go]: What's this condition mean?

File: store/localstore/local_region.go [hunk 9]
```diff
@@ -505,7 +689,7 @@ func (rs *localRegion) getRowsFromIndexReq(ctx *selectContext) error {
 		if limit == 0 {
 			break
 		}
-		count, err := rs.getIndexRowFromRange(ctx, ran, desc, limit)
+		count, err := rs.getIndexRowFromRange(ctx, ran, ctx.descScan, limit)
```
Comment (shenli) [inline store/localstore/local_region.go]: Can we save the third parameter?

File: store/localstore/local_region.go [hunk 10]
```diff
@@ -382,6 +538,31 @@ func (rs *localRegion) handleRowData(ctx *selectContext, handle int64, value []b
 	return rs.valuesToRow(ctx, handle, values)
 }
 
+func (rs *localRegion) evalTopN(ctx *selectContext, handle int64, values map[int64][]byte, columns []*tipb.ColumnInfo) error {
```
Comment (shenli) [inline store/localstore/local_region.go]: Add comment for this function.

File: store/localstore/local_region.go (hunk 8 above)
Comment (hanfei1991) [inline store/localstore/local_region.go]: If orderby[0].expr is nil and sel.orderby[0].desc = false, it means this request is desc scanned.

File: store/localstore/local_region.go (hunk 9 above)
Comment (hanfei1991) [inline store/localstore/local_region.go]: function "getRowsFromRange" also needs this var. So it's better to calculate it outside.

**Comment by shenli** [comment]
LGTM

File: store/localstore/local_region.go (hunk 8 above)
Comment (BusyJay) [inline store/localstore/local_region.go]: Why desc = false means doing desc scanning?

Comment (hanfei1991) [inline store/localstore/local_region.go]: desc = true... I made a mistake.

**Pull request merged**